src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,112 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int START_CAPACITY = 10;
+    private T[] values;
+    private int size;
+
+    public ArrayList() {
+        values = (T[]) new Object[START_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        addFreeSpace(values.length);
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+        } else {
+            checkIndex(index);
+            addFreeSpace(values.length);
+            T[] newValues = (T[]) new Object[values.length];
+            for (int i = 0; i < index; i++) {
+                newValues[i] = values[i];
+            }
+            newValues[index] = value;
+            for (int i = index; i < size; i++) {
+                newValues[i + 1] = values[i];
+            }
+            size++;
+            values = newValues;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T[] newValues = (T[]) new Object[values.length];
```

Comment:
let's use the same array in System.arraycopy method
just shift values in the same array